frame/proxy/src/lib.rs

```diff
@@ -108,19 +107,67 @@ pub trait Trait: frame_system::Trait {
 
 	/// Weight information for extrinsics in this pallet.
 	type WeightInfo: WeightInfo;
+
+	/// The maximum amount of time-delayed announcements that are allowed to be pending.
+	type MaxPending: Get<u32>;
+
+	/// The type of hash used for hashing the call.
+	type CallHasher: Hash;
+
+	/// The base amount of currency needed to reserve for creating an announcement.
+	///
+	/// This is held when a new storage item holding a `Balance` is created (typically 16 bytes).
+	type AnnouncementDepositBase: Get<BalanceOf<Self>>;
+
+	/// The amount of currency needed per announcement made.
+	///
+	/// This is held for adding an `AccountId`, `Hash` and `BlockNumber` (typically 68 bytes)
+	/// into a pre-existing storage value.
+	type AnnouncementDepositFactor: Get<BalanceOf<Self>>;
 }
 
+/// The parameters under which a particular account has a proxy relationship with some other
+/// account.
+#[derive(Encode, Decode, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, RuntimeDebug)]
+pub struct ProxyDefinition<AccountId, ProxyType, BlockNumber> {
+	/// The account which may act on behalf of another.
+	delegate: AccountId,
+	/// A value defining the subset of calls that it is allowed to make.
+	proxy_type: ProxyType,
+	/// The number of blocks that an announcement must be in place for before the corresponding call
+	/// may be dispatched. If zero, then no announcement is needed.
+	delay: BlockNumber,
+}
+
+/// Details surrounding a specific instance of an announcement to make a call.
+#[derive(Encode, Decode, Clone, Copy, Eq, PartialEq, RuntimeDebug)]
+pub struct Announcement<AccountId, Hash, BlockNumber> {
+	/// The account which made the announcement.
+	real: AccountId,
+	/// The hash of the call to be made.
+	call_hash: Hash,
+	/// The height at which the announcement was made.
+	height: BlockNumber,
+}
+
+type CallHashOf<T> = <<T as Trait>::CallHasher as Hash>::Output;
+
 decl_storage! {
 	trait Store for Module<T: Trait> as Proxy {
 		/// The set of account proxies. Maps the account which has delegated to the accounts
 		/// which are being delegated to, together with the amount held on deposit.
-		pub Proxies: map hasher(twox_64_concat) T::AccountId => (Vec<(T::AccountId, T::ProxyType)>, BalanceOf<T>);
+		pub Proxies: map hasher(twox_64_concat) T::AccountId
+			=> (Vec<ProxyDefinition<T::AccountId, T::ProxyType, T::BlockNumber>>, BalanceOf<T>);
+
+		/// The announcements made by the proxy (key).
+		pub Announcements: map hasher(twox_64_concat) T::AccountId
+			=> (BalanceOf<T>, Vec<Announcement<T::AccountId, CallHashOf<T>, T::BlockNumber>>);
```

Comment:
ocd nit, balance the second value in the tuple in one case, and the first value in another